docs/registry.md

```diff
@@ -0,0 +1,274 @@
+# Registry
+
+The registry is a place where applications metadata are stored and versioned. It can be used by a cozy 
+
+We define an application registry as an API. This should allow us to defer the real implementation of the registry storage and allow different store implementations.
+
+
+## Objects
+
+Two type of objects are managed in the registry: applications and versions.
+
+### Application
+
+An application object contains the following fields:
+
+- name: the application name
+- editor: the application editor
+- category: the application category
+- repository: object with type and url of package repository
+- license: the application license
+- versions: a list of all versions
+
+### Version
+
+An application version object contains the following fields:
+
+- name: the application name
+- description: description from the package.json
+- version: the version string
+- url: url of the tarball containing the application at specified version
+- sha256: the sha256 checksum of the application content
+- permissions: the permissions map contained in the manifest
+
+## APIs: Adding to registry
+
+These APIs can be used to add elements to the registry.
+
+### POST /:app
+
+This route adds an application to the registry. The content of the request should be a json object of an application.
+
+#### Status codes
+
+* 201 Created, when the application has been successfully added
+* 409 Conflict, when an application with the same name already exists
+* 400 Bad request, if the given application data is malformed (bad name, missing editor, ...)
+
+#### Request
+
+```http
+POST /drive
+
+{
+    "name": "drive",
+    "editor": "cozy",
+    "description": "The drive application",
+    "versions": ["v3.1.1","v3.1.2"],
+    "repository": "https://github.com/cozy/cozy-drive",
+    "license": "BSD"
+}
+```
+
+### POST /:app/:version
+
+This route adds a version of an application to the registry. The application content can be transmitted by two different mean:
+
+    - via a `url` query parameter specifying the url of the gzipped tarball
+    - via the content of the POST request in which case the registry should store the application on its own.
+
+The content of the manifest file is used to fill the fields of the version.
+
+Before adding the application version to the registry, the registry should check the following:
+
+    - the `manifest` file contained in the tarball should be checked and have its fields checked against the application properties
+    - the application content should be checked against the given `sha256` query parameter
```

Comment:
May be some text here about how most registry (including cozy's) will probably add security auto/manual checks  and therefore the application may or may not be available immediately upon creation.